rust/noosphere/src/ffi/petname.rs

```diff
@@ -1,47 +1,173 @@
 #![allow(unused_variables)]
 
-use safer_ffi::prelude::*;
+use std::str::FromStr;
 
-use crate::ffi::NsError;
+use anyhow::anyhow;
+use cid::Cid;
+use noosphere_sphere::{SpherePetnameRead, SpherePetnameWrite, SphereWalker};
+use safer_ffi::{char_p::InvalidNulTerminator, prelude::*};
 
-use super::NsNoosphereContext;
+use crate::ffi::{NsError, TryOrInitialize};
+
+use super::{NsNoosphere, NsSphere};
 
 #[ffi_export]
-/// Get the DID that is assigned to the provided petname; note that the DID is
-/// the ID of the sphere, but in order to read the sphere you must resolve the
-/// DID to a CID, which tells you the version of the sphere to read.
+/// Get the sphere identity - a DID - assigned to a given petname of a sphere.
+/// Note that this call will produce an error if no address has been assigned to
+/// the name.
 pub fn ns_sphere_petname_get(
-    noosphere: &NsNoosphereContext,
-    sphere_identity: char_p::Ref<'_>,
+    noosphere: &NsNoosphere,
+    sphere: &NsSphere,
     petname: char_p::Ref<'_>,
     error_out: Option<Out<'_, repr_c::Box<NsError>>>,
 ) -> Option<char_p::Box> {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner()
+                .get_petname(petname.to_str())
+                .await?
+                .ok_or_else(|| anyhow!("No petname '{}' has been set", petname.to_str()))?
+                .to_string()
+                .try_into()
+                .map_err(|error: InvalidNulTerminator<String>| anyhow!(error).into())
+        })
+    })
 }
 
 #[ffi_export]
-/// Assign a DID to a petname. This will overwrite a petname entry if one already exists
-/// with the given name (and reset the resolved CID, if any).
+/// Assign a petname to a sphere identity (a DID). This will overwrite the
+/// petname with the new sphere identity if it has been set previously. When a
+/// new address is assigned to a petname, the petname will be reset to an
+/// unresolved state. Make sure to invoke sync after setting new petnames, so
```

Comment:
What happens if an "unresolved state" is propagated in a sphere? This reads as "undefined behaviour"

Reply:
A name / DID pair resolve to a link, just like in DNS a domain name resolves to the value side of a DNS record. If it is resolved, you have a link. If it is unresolved, you don't have a link.